### ISSUE TITLE
false positive for warning 26 with some mll files

### ISSUE THREAD

**Issue by redianthus** [ocaml/ocaml#10076]
Hi,

I found this issue on [this repository](https://github.com/ocaml-toml/To.ml/commit/e63dac763996b73532480092f94d6ebd504234e8).
If you remove [this line](https://github.com/ocaml-toml/To.ml/blob/e63dac763996b73532480092f94d6ebd504234e8/src/dune#L6) (which is simply the line disabling warning 26) and run `dune build` you'll get the following :

```
File "src/tomlLexer.mll", line 20, characters 27-35:
Error (warning 26): unused variable int_sign.
File "src/tomlLexer.mll", line 20, characters 53-61:
Error (warning 26): unused variable int_part.
```

They are defined on [this line](https://github.com/ocaml-toml/To.ml/blob/e63dac763996b73532480092f94d6ebd504234e8/src/tomlLexer.mll#L20).

But this is wrong as they're clearly used [here](https://github.com/ocaml-toml/To.ml/blob/e63dac763996b73532480092f94d6ebd504234e8/src/tomlLexer.mll#L46) and [there](https://github.com/ocaml-toml/To.ml/blob/e63dac763996b73532480092f94d6ebd504234e8/src/tomlLexer.mll#L47).

**Comment by dra27** [comment]
The compiler is correct - the location forced by ocamllex is misleading. You have this definition:

```ocaml
let t_int     = (t_sign as int_sign)? (t_int_part as int_part)
```

which you use in this rule:

```ocaml
  | t_int as value   {
      let int_string =
        match int_sign with
        | Some '+'  -> int_part
        | _         -> value
      in INTEGER (int_of_string int_string)
    }
```
and which indeed does use the bound values `int_sign` and `int_part`. However, you also have this definition:

```ocaml
let t_float   = t_sign? t_int_part ((t_frac t_exp?) | t_exp)
```
which is used in this rule:
```ocaml
  | t_float as value   { FLOAT (float_of_string value) }
```
which does not use the two bound values and which is responsible for the warning.

It's fairly trivial to alter ocamllex to report the location of the use rather than the definition. It's not perfect, but I think it's better to point to the actual usage, even though doesn't point to the name itself, rather than pointing to the strict definition of the name and having no idea where it came from!

**Comment by lpw25** [comment]
I disagree with this assessment of the issue. Both those bindings are used somewhere in the file: if you remove them the file will stop working. We don't want to prevent people from sharing patterns just because they only sometimes need the binding. Code generators like ocamllex should just be disabling warnings on their produced code.

**Comment by redianthus** [comment]
Still an issue.

**Comment by gasche** [comment]
@zapashcanon would you be willing to try to fix this in OCamllex?

This could be done with explicit `ignore x;` inserted in the generated code, or appropriate `[@@ocaml.warning ...]` directives. The directives could be added globally on the whole file (ignore all unused variables), or locally on variables bound by regexp patterns (better, but maybe more work).

**Comment by redianthus** [comment]
@gasche fine, I'll have a look. Can you assign me so I don't forget to do it ?

**Comment by gasche** [comment]
(I'm still hoping that @zapashcanon would eventually look at it.)

**Comment by redianthus** [comment]
(I still hope to find the time!)
